File: linked-list/linked_list.rb
class Node
  attr_accessor :data, :next
  def initialize(data)
    @data, @next = data, nil
  end
end

class LinkedList
  def initialize
    @head = nil
  end
  
  def append(data)
    new_node = Node.new(data)
    if @head.nil?
      @head = new_node
    else
      current = @head
      current = current.next while current.next
      current.next = new_node
    end
  end
  
  def display
    current = @head
    while current
      print "#{current.data} -> "
      current = current.next
    end
    puts "nil"
  end
end

ll = LinkedList.new
[1, 2, 3].each { |x| ll.append(x) }
ll.display

File: linked-list/linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* next;
} Node;

typedef struct {
    Node* head;
} LinkedList;

LinkedList* createList() {
    LinkedList* list = malloc(sizeof(LinkedList));
    list->head = NULL;
    return list;
}

void append(LinkedList* list, int data) {
    Node* newNode = malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;
    
    if (!list->head) {
        list->head = newNode;
        return;
    }
    
    Node* current = list->head;
    while (current->next) current = current->next;
    current->next = newNode;
}

void display(LinkedList* list) {
    Node* current = list->head;
    while (current) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

int main() {
    LinkedList* list = createList();
    append(list, 1);
    append(list, 2);
    append(list, 3);
    display(list);
    return 0;
}